hexa/files/backends/fs.py

```diff
@@ -0,0 +1,325 @@
+
+    def bucket_exists(self, bucket_name: str):
+        return self.exists(bucket_name)
+
+    def exists(self, name):
+        try:
+            exists = os.path.lexists(self.path(name))
+            return exists
+        except self.exceptions.SuspiciousFileOperation:
+            raise
+
+    def path(self, *paths):
+        return safe_join(self.data_dir, *paths)
+
+    def size(self, name):
+        return os.path.getsize(name)
+
+    def _get_payload_from_token(self, token):
+        try:
+            signer = TimestampSigner()
+            decoded_token = force_str(urlsafe_base64_decode(token))
+            payload = signer.unsign_object(decoded_token, max_age=self._token_max_age)
+            return payload
+        except (UnicodeDecodeError, BadSignature, ValueError):
+            raise self.exceptions.BadRequest("Invalid token")
+
+    def _create_token_for_payload(self, payload: dict):
+        signer = TimestampSigner()
+        signed_payload = signer.sign_object(payload, compress=True)
+        return urlsafe_base64_encode(force_bytes(signed_payload))
+
+    def to_storage_object(self, bucket_name: str, object_key: Path):
+        full_path = self.path(bucket_name, object_key)
+        if not self.exists(full_path):
+            raise self.exceptions.NotFound(f"Object {object_key} not found")
+        if full_path.is_file():
+            return StorageObject(
+                name=object_key.name,
+                key=object_key,
+                updated=datetime.fromtimestamp(
+                    os.path.getmtime(str(full_path))
+                ).isoformat(),
+                size=self.size(full_path),
+                path=Path(bucket_name) / object_key,
+                type="file",
+                content_type=guess_type(full_path)[0] or "application/octet-stream",
+            )
+        else:
+            return StorageObject(
+                name=object_key.name,
+                key=object_key,
+                updated=datetime.fromtimestamp(os.path.getmtime(str(full_path))),
+                path=Path(bucket_name) / object_key,
+                type="directory",
+            )
+
+    def get_valid_filepath(self, path: str | Path):
+        """Returns a path where all the directories and the filename are valid.
+
+        Args:
+            path (str|Path): A path
+        """
+        return "/".join(get_valid_filename(part) for part in str(path).split("/"))
+
+    def create_directory(self, directory_path: str):
+        if self.exists(directory_path):
+            raise self.exceptions.AlreadyExists("Directory already exists")
+        directory = self.path(directory_path)
+
+        try:
+            if self.directory_permissions_mode is not None:
+                # Set the umask because os.makedirs() doesn't apply the "mode"
+                # argument to intermediate-level directories.
+                old_umask = os.umask(0o777 & ~self.directory_permissions_mode)
```

Comment:
thank you, interesting!

I wonder if  here instead os.mkdirs() - path.mkdir(parents=True) is better to be explicit for the parent folder generation?

Reply:
By looking at the documentation of both calls:
- https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir
- https://docs.python.org/3/library/os.html#os.makedirs

Only the `os.makedirs` proposes a way to set the permissions to intermediate-directories